### ISSUE TITLE
Export the trained model ?

### ISSUE THREAD

**Issue by behrica** [asreview/asreview#283]
If we would use AI in our systematic reviews,
"transparency" would be very important.

This might mean that, if I do and finalize a review "today",
I want to be able to "save" and archive everything, which will ideally allow me to 
"reproduce" the results of the AI in one year, 5 years, 10 years..

This means to me, at least, that we can "access and archive" the trained model in some form,
which should at least allow me to answer the questions like:

"would this paper have been included by your model ?"

Of course, I needed to save more then the "trained model",  ( like libraries)
but the model would be the start.

**Comment by qubixes** [comment]
@behrica At the risk of sounding pedantic, in our current review design, only the reviewer includes papers.

Maybe the following is a question you would like to ask?

"Given a paper not included in the review, how high would it score compared to papers in the dataset, given the state of the review at time X?"

In that case, I agree that your question can be answered by storing the trained model in some form or another for each training event (which can be many!). Note that scores/probabilities are already stored for each record at each training event.

As you might have noticed already, providing full reproducibility is going to be difficult, since it depends on timing (user/system).

The thing that would be the easiest to implement would be to store the seed for the training step, but I'm not sure if that satisfies your requirement?

**Comment by behrica** [comment]
> @behrica At the risk of sounding pedantic, in our current review design, only the reviewer includes papers.
> 
> Maybe the following is a question you would like to ask?
> 
> "Given a paper not included in the review, how high would it score compared to papers in the dataset, given the state of the review at time X?"
Yes, that sounds good.

I think what at then end we want to be able to do, with an AI in systematic review is to have an answer in the following scenario:

1. We do an SR, starting from 1000 papers from a certain literature search
2. Experts and AI pick 30 to be included
3. We produce a output based on the 30 picked
4. One year later(or 10 years later)  somebody comes with paper 1001, claiming that we forgot to look at it in the past (the search missed it, for example)
5. Now we want to be able to tell, if the model we used in the past would have included or excluded the paper
6. If "excluded" we are fine. If included we might need to "reopen" our output, if there is different evidence in the paper
Specifically, an answer like "we don't know" what would have happend, is not be acceptable.


We are piloting as well to address screening as an MLT "text classification" probelm. And in there, we could in theory store the trained model and run it, one year later, on the new paper and we get an answer "excluded" or "included". From this comes the question to export and store the trained model.
We think as well about "export" here, as we want to shield ourselves from unpredicted changes in the review application. So ideally we can answer this question as well in 10 years after the SR was done.

I am rather new to the concept of "active learning" as done by ASreview, so maybe I just misunderstand something.

Point 5) does not need to be a feature of ASReview, because it is a very rare case. But we would like to have archived all pieces, so that we COULD at least do it. Again, a classification model I could just store in archive.

**Comment by behrica** [comment]
I just looked at the code. It is based on sklearn, as I saw.

So the think to export, store, would be the "pipeline" object,
containing the fitted preprocessors and the fitted models.

This can be "pickled" to a file and archived.

The, in theory, unpickling this and having the same python + library versions,
I should be able to call   xxx.fit("new paper text") on it and it would tell me "included", "excluded"

**Comment by qubixes** [comment]
In the active learning model we use, the AI only _proposes_ papers, and the expert _decides_ whether those papers should be included. That is why I think the fitted model at the end of the review is not particularly useful, since no inclusions will be made after the expert stops reviewing.

So I'd state it as:

5*. Would ASReview have proposed the new paper during the review?

For this question you would need the pipeline object + fitted model at every time step.

Personally, I would rather see the AI assisted SR as a statistical process (because it is!) and ask the question:

5**: What is the chance that the new paper would have been proposed?

This question could be answered by only storing the exact python/libraries + state file. But of course I don't know if that would be acceptable in your line of work. (I would argue the answer to this question contains more information than an answer to 5*/5.)

**Comment by behrica** [comment]
5** Is fine for me.

Can you explain why for this, we don't need the "serialized pipeline/model ?

Maybe because the model can be re-trained / re-created from state file (+ libraries  + initial papers ?)

**Comment by qubixes** [comment]
Yes, if you have the state file + dataset file + libraries, that should be enough to retrain the model as if you were doing the review. The state file contains the model + parameters + initial papers, etc. It also contains the information on which papers the model was trained on at each training step.

So you could compute the probability of inclusion at each (training) junction of not finding the new paper, and multiply them all up to get the probability of failing to find the new paper during the review. I'm not sure if that would be optimal/how to compute the error, but I think that at least asymptotically that should be correct.

**Comment by behrica** [comment]
Thanks for the detailed answer. We can close the issue here.
